jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -71,6 +71,9 @@
 	static final boolean enableCompression = com.ibm.oti.vm.VM.J9_STRING_COMPRESSION_ENABLED;
 	
 	/*[IF Sidecar19-SE]*/
+	/*[IF Java18.3]*/
+	private final byte coder = coder();
```

Comment:
What is this for? It appears unused.

Reply:
We should be careful adding instance fields to string due to the footprint increase they cause

Reply:
This field is required for Java 18.3 due to references by j.i.UnixFileSystem.canonicalize0.

Reply:
As Andrew pointed out this will (negatively) affect footprint: Perhaps we ought to fix UnixFileSystem.canonicalize0?

Reply:
`j.i.UnixFileSystem.canonicalize0` is part of `OpenJDK`. The general idea is to avoid JCL patch as much as possible. 
`Java 18.3` code is still progressing so it might be changed in future builds. In case it stays in final release, the footprint impact on `Java 18.3` due to this additional instance field could be addressed later.

Reply:
I don't see the word 'coder' in https://github.com/AdoptOpenJDK/openjdk-jdk10/blob/master/jdk/src/java.base/unix/classes/java/io/UnixFileSystem.java or https://github.com/AdoptOpenJDK/openjdk-jdk10/blob/master/jdk/src/java.base/unix/native/libjava/UnixFileSystem_md.c: if the reference is already gone, I'd prefer not to add the field to String.

Reply:
Not sure about *double* part, there are a few other instance fields within `j.l.String` such as 
	`private final byte[] value;`
	`private final int count;`
	`private int hashCode;`

`Java 18.3` `b23` -version throws following exception without the `coder` field
````Exception in thread "main" java/lang/NoSuchFieldError: java/lang/String.coder
        at java/io/UnixFileSystem.canonicalize0 (java.base@9/NativeMethod:4294967295)
        at java/io/UnixFileSystem.canonicalize (java.base@9/UnixFileSystem.java:170)
        at java/io/File.getCanonicalPath (java.base@9/File.java:618)
        at java/io/File.getCanonicalFile (java.base@9/File.java:643)
        at jdk/internal/loader/URLClassPath.toFileURL (java.base@9/URLClassPath.java:253)
        at jdk/internal/loader/URLClassPath.<init> (java.base@9/URLClassPath.java:188)
        at jdk/internal/loader/ClassLoaders.<clinit> (java.base@9/ClassLoaders.java:85)
        at java/lang/ClassLoader.initializeClassLoaders (java.base@9/ClassLoader.java:183)
        at java/lang/Thread.initialize (java.base@9/Thread.java:422)
        at java/lang/Thread.<init> (java.base@9/Thread.java:153)````

Reply:
OpenJDK `Java 18.3` only has following three instance fields:
  `private final byte[] value;`
  `private final byte coder;`
  `private int hash;`
There is no `count`.
fyi @fjeremic

Reply:
We bundle to `coder` and the `count` into a single field thus getting the benefit of having a `count` which speeds up substring operations and StringBuilder interaction while also supporting string compression.

Reply:
Adding the coder increases footprint of all strings in compressedrefs by 33% (currently 3 fields so `J9Class*` plus 3 32-bit fields is 16 bytes. Adding another field bumps this to 20, but because of alignment it increases to 24. 1 - 16/24 = 33%. This is definitely something we don't want.

Reply:
I think it's worse: my understanding is that the unit of allocation is now 16 bytes (so everything is 16-byte aligned). String is 16 bytes in Java 8 & 9, the extra byte requires another 16 bytes for a total of 32 bytes.

Reply:
I feel OpenJ9 should have a field `coder` to match OpenJDK for two reasons:
1. OpenJ9 with same `j.l.String` fields as OpenJDK can avoid JCL patches required to fix `NoSuchFieldError` as per earlier comment in this thread. Avoiding JCL patch is as important as performance advantages. Past experience shows JCL patch price tag is high in all perspectives;
2. User applications attempting reflection access of `j.l.String` field `coder` will succeed in OpenJDK but fail at OpenJ9 which usually causes confusion and defect against J9.

Can field `coder` serve the purpose of OpenJ9 only field `count`? I suggest to open another issue for further discussion to rename the field `count` to `coder`. This pull request (adding field `coder`) won't affect `Java 8 & 9`, and allow adoption of `Java 18.3` further updates.

Reply:
> I think it's worse: my understanding is that the unit of allocation is now 16 bytes (so everything is 16-byte aligned). String is 16 bytes in Java 8 & 9, the extra byte requires another 16 bytes for a total of 32 bytes.

You're right! We support compressedrefs shift of 4 so 2^4 = 16 byte alignment. This change will double the size of all string objects.

Reply:
The coder field does not serve the same purpose as count. There is an important difference in the String implementation behavior here for OpenJ9. In OpenJ9 the backing array of a string can be longer than the string length. This is a very useful optimization for a number of cases including when a StringBuilder/StringBuffer has append called after a toString; a substring is used to generate prefix strings; and to prevent having to truncate the backing array of a StringBuilder/StringBuffer when toString is called if the builder's of buffer's backing array is over allocated. Without the count field these optimizations cannot be done. Removing or repurposing the count field does have a major performance implication as a result.

Adding a new field does need a good justification given the huge object size increases that can be seen (per the discussion above). I agree there is a case for minimizing patches, but given how performance sensitive String is a little bit more study might be justified before adopting the field.

Reply:
The only with renaming the field from count to coder is that count is an int and coder field is a byte. The byte may only hold 0 or 1 or something like that, but for count the MSB is the bit so I don't know that we can make the reflective access equivalent...

Reply:
Related to https://github.com/eclipse/openj9/issues/223.

Reply:
We would go ahead and deliver this change to temporarily add the coder field. As @JasonFengJ9 said, the addition of the coder field only affects Java 18.3, for which there isn't yet any OpenJ9 build provided. The extensions repo for Java 18.3 hasn't been created yet, so we don't have a simple way to patch the JCL code at this time. It can be done later after the extensions repo is created, #223 is opened to track this. Delivering this change enables raw builds of Java 18.3, which is desirable in order to provide initial support for Java 18.3 and make it easier to stay up to date and keep it working.

Reply:
Note that the coder field will likely not have the correct value once OpenJ9 enables string compression by default leading the presence of this field to cause incorrect behavior.

Reply:
Does coder() return correct coder value, or the coder value might be changed after string construction?

Reply:
coder() does so the initialization was right - I take back my comment it should be correct.